### ISSUE TITLE
option_if_let_else doesn't seem to handle conditions on match patterns.

### ISSUE THREAD

**Issue by tdkt** [rust-lang/rust-clippy#9742]
### Summary

This lint generates a suggestion that has a different logic from the original code. Probably this lint should not fire when the pattern isn't simple.

### Reproducer

I tried this code:

```rust
#![warn(clippy::nursery)]

fn main()
{
	let _a = problem();
}


struct Spec
{
	name: Option<&'static str>,
}

impl Spec
{
	fn validate(&self) -> Option<&'static str>
	{
		Some("validated")
	}
}

fn problem() -> Option<&'static str>
{
	let spec = Spec{ name: Some("test") };
	let method = "test2";

	match spec.name
	{
		Some(name) if name == method => spec.validate(),
		_ => None,
	}
}

```

I expected to see this happen:

Clippy to be happy. 
Instead, this happened:

```
warning: use Option::map_or instead of an if let/else
  --> src/main.rs:27:5
   |
27 |       match spec.name
   |  ________^
28 | |     {
29 | |         Some(name) if name == method => spec.validate(),
30 | |         _ => None,
31 | |     }
   | |_____^ help: try: `spec.name.map_or(None, |name| spec.validate())`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#option_if_let_else
   = note: `#[warn(clippy::option_if_let_else)]` implied by `#[warn(clippy::nursery)]`
```

### Version

```text
rustc 1.66.0-nightly (0da281b60 2022-10-27)
binary: rustc
commit-hash: 0da281b6068a7d889ae89a9bd8991284cc9b7535
commit-date: 2022-10-27
host: aarch64-apple-darwin
release: 1.66.0-nightly
LLVM version: 15.0.2
```

### Additional Labels

@rustbot label +`l-suggestion-causes-error`

**Comment by tdkt** [comment]
@kraktus Thank you! Your bug smashing is swift and fierce ;)
